tests - massive test refactor (#106)
ansible-collections/community.hashi_vault: docs/docsite/rst/contributor_guide.rst

.. _ansible_collections.community.hashi_vault.docsite.contributor_guide:

Contributor Guide

> =================

Author: Thank you so much for this link! I've re-done the headers on all three pages to (hopefully) match. That also inspired lots of other formatting changes to make the docs fit better, and I tried to apply more of the style guide recommendations on header style (Sentence case), language style, etc.

I also added an entire new section specifically on contributing documentation, and updated all the links I could to `:ref:`erences (this took a fair bit of trial and error, the doc build workaround from Sam was essential for me to test all this out).

Skip to the next section for a method that's nearly as easy but better off in the long run (docker-compose).

Legacy Mode
...........

To get started quickly without having to set anything else, you can use legacy mode by copying the included integration config sample:

.. code-block:: bash

    $ cp tests/integration/integration_config.yml.sample tests/integration/integration_config.yml

That file has everything configured to be able to run the integration tests and have them set up the dependencies for you.

You will also need the following additional Ansible collections:

* `community.crypto <https://galaxy.ansible.com/community/crypto>`_
* `community.general <https://galaxy.ansible.com/community/general>`_ (MacOS local/venv only)

Running legacy mode tests in docker (recommended):

.. code-block:: bash

    $ ansible-test integration --docker default -v

Running legacy mode tests in a controlled python virtual environment (**not recommended**):

.. code-block:: bash

    $ ansible-test integration --venv --requirements --allow-destructive -v

Note that your system packages may be manipulated by running locally or in a venv.

> Legacy mode is not recommended because a new Vault server and proxy server will be downloaded, set up, configured, and/or uninstalled, for every *target*. Traditionally, we've only had one target, and so it was a good way to go, but that's no longer going to be the case. This is going to make it slower and slower as you'll incur the overhead on every target, in every run.

Reviewer: This is a really convincing argument. I'd move it up to the beginning of the section, so folks might read it before they follow the instructions on running tests in legacy mode.

Author: Perfecto 👌 moved into a nice note box, slightly re-worded, thank you!

> As a result, if you're going to use legacy mode it's best to limit your test run to the specific target needed, for example:

Reviewer: If you do move the paragraph above, then this line would need to change - maybe something like:
```suggestion
If you must use legacy mode testing, you can make it more efficient by limiting your test run to the specific target needed, for example:
```

.. _ansible_collections.community.hashi_vault.docsite.contributor_guide:

*****************

> Contributor guide

Reviewer: Hm, the GitHub standard is to add a `contributing.md` file, right? Maybe add one and point to this file or to the built page?

Author: I will leave `contributing.md` for a later time, but I have updated the (previously blank) section in the README.

*****************

This guide aims to help PR authors contribute to the ``community.hashi_vault`` collection.

.. note::

> 	This guide is a work-in-progress and should not be considered complete. Check back often as we fill out more details based on experience and feedback, and please let us know how this guide can be improved.

Author: Yes exactly, added that (with link)

.. contents::
  :local:
  :depth: 3

Quick start
===========

#. Log into your GitHub account.
#. Fork the `ansible-collections/community.hashi_vault repository <https://github.com/ansible-collections/community.hashi_vault>`_ by clicking the **Fork** button in the upper right corner. This will create a fork in your own account.

> #. Clone the repository locally, following :ref:`the example instructions here <hacking_collections>` (but replace ``general`` with ``hashi_vault``). **Pay special attention to the path structure.**

Author: Updated!

#. As mentioned on that page, commit your changes to a branch, push them to your fork, and create a pull request (GitHub will automatically prompt you to do so when you look at your repository).
#. :ref:`See the guidance on Changelogs <community_changelogs>` and include a :ref:`changelog fragment <changelogs_how_to>` if appropriate.

Contributing documentation
==========================

Additions to the collection documentation are very welcome! We have three primary types of documentation, each with their own syntax and rules.

Module and plugin documentation
-------------------------------

This type of documentation gets generated from structured YAML, inside of a Python string. It is included in the same code that it's documenting, or in a separate Python file, such as a doc fragment. Please see the :ref:`module format and documentation guidance <developing_modules_documenting>` for more information.

This type of documentation is tested with ``ansible-test sanity`` and full instructions are available on the :ref:`testing module documentation <testing_module_documentation>` page.

> Although we can't preview how the documentation will look for these, we can be reasonably assured the output is correct because of the highly structured nature and the validation via sanity tests.

Reviewer: ```suggestion
Although we can't preview how the documentation will look for these, we can be reasonably sure the output is correct because the documentation is highly structured validated using sanity tests.
```

Reviewer: ugh, I deleted the `and` there by mistake . . . "highly structured and validated using . . . "

Author: oh nice, I can update your suggestion and then commit it

Only the ``extra-docs.yml`` file is validated by the collection's CI, and there is not yet any automated preview for the docsite rendering, but this is an area we are hoping to improve on.

Advanced
^^^^^^^^

Docsite pages can be generated locally through a workaround technique. **This is not a supported method** but it may be helpful to get more rapid feedback on docsite if changes, if you're comfortable at a command line.

The process is:

#. Clone `ansible/ansible <https://github.com/ansible/ansible/>`_ or a fork of it.
#. Copy the ``.rst`` files you want to preview into that repository's ``docs/docsite/`` directory.
#. Install the requirements needed to build the docsite (from the repository root):

    .. code-block:: shell-session

      $ pip install -r requirements.txt
      $ pip install -r docs/docsite/requirements.txt

#. You may also need to remove write permission from group and other:

    .. code-block:: shell-session

      $ chmod -R go-w docs/docsite/rst

#. Build the docs:

    .. code-block:: shell-session

      $ make coredocs

The rendered HTML docs should be available in ``docs/docsite/_build/html/`` and can be opened in a browser.

> Running Tests Locally

Author: This is not related to documentation anymore; I thought it being a peer header to the documentation section marked it enough. It looks ok to me rendered, especially in the TOC
![image](https://user-images.githubusercontent.com/1260690/126229053-8d7d249e-d968-46dc-90f5-c5b80aa98a15.png)

=====================

> If you're making anything more than very small or one-time changes, you'll want to run the tests locally to avoid having to push a commit for each thing, and waiting for the CI to run tests.

Reviewer: Sneaky future tense here:
```suggestion
If you're making anything more than very small or one-time changes, run the tests locally to avoid having to push a commit for each thing, and waiting for the CI to run tests.
```
